lib/grovepi/board.ex

```diff
@@ -56,40 +60,112 @@ defmodule GrovePi.Board do
   not normally called directly.
   """
   @spec get_response(atom, integer) :: binary | {:error, term}
-  def get_response(prefix, len) do
-    get_response_with_retry(i2c_name(prefix), len, @i2c_retry_count)
+  def get_response(prefix, bytes_to_read) do
+    GenServer.call(i2c_name(prefix), {:read, bytes_to_read})
   end
 
   @spec get_response(integer) :: binary | {:error, term}
-  def get_response(len) do
-    get_response(Default, len)
+  def get_response(bytes_to_read) do
+    get_response(Default, bytes_to_read)
   end
 
   @doc """
   Write directly to a device on the I2C bus. This is used for sensors
   that are not controlled by the GrovePi's microcontroller.
   """
-  def i2c_write_device(address, buffer) do
-    @i2c.write_device(i2c_name(Default), address, buffer)
+  def i2c_write_device(address, message) do
+    GenServer.call(i2c_name(Default), {:write_device, address, message})
+  end
+
+  #### test helper functions
+
+  def add_responses(board, messages) do
+    GenServer.call(board, {:add_responses, messages})
+  end
+
+  def add_response(board, message) do
+    add_responses(board, [message])
+  end
+
+  def get_last_write(board) do
+    GenServer.call(board, {:get_last_write})
+  end
+
+  def get_last_write_data(board) do
+    GenServer.call(board, {:get_last_write_data})
+  end
+
+  def get_all_writes(board) do
+    GenServer.call(board, {:get_all_writes})
+  end
+
+  def get_all_data(board) do
+    GenServer.call(board, {:get_all_data})
+  end
+
+  def reset(board) do
+    GenServer.call(board, :reset)
   end
 
-  # The GrovePi has intermittent I2C communication failures. These
-  # are usually harmless, so automatically retry.
-  defp send_request_with_retry(_board, _message, 0), do: {:error, :too_many_retries}
+  ## Server Callbacks
 
-  defp send_request_with_retry(board, message, retries_left) do
-    case @i2c.write(board, message) do
-      {:error, _} -> send_request_with_retry(board, message, retries_left - 1)
-      response -> response
-    end
+  @impl true
+  def init(state) do
+    {:ok, state, {:continue, []}}
   end
 
-  defp get_response_with_retry(_board, _len, 0), do: {:error, :too_many_retries}
+  @impl true
+  def handle_continue(_continue, state) do
+    {:ok, ref} = @i2c.open("i2c-1")
+    {:noreply, %{state | i2c_bus: ref}}
+  end
+
+  @impl true
+  def handle_call({:write, message}, _from, state) do
+    reply = @i2c.write(state.i2c_bus, state.address, message, retries: @i2c_retry_count)
+    {:reply, reply, state}
+  end
+
+  @impl true
+  def handle_call({:write_device, address, message}, _from, state) do
+    reply = @i2c.write(state.i2c_bus, address, message, retries: @i2c_retry_count)
+    {:reply, reply, state}
+  end
+
+  @impl true
+  def handle_call({:read, bytes_to_read}, _from, state) do
+    reply =
+      case(@i2c.read(state.i2c_bus, state.address, bytes_to_read, retries: @i2c_retry_count)) do
+        {:ok, response} -> response
```

Comment:
I think that you want to keep the `{:ok, _}` part of the tuple here so that callers can more easily detect success.

Reply:
Looks like this was the old behavior, so I take back my comment for this PR.